Restrict releases by Location

mkangia opened the pull request:
[Sneak Preview](http://recordit.co/8PcgUyR7ZT)

[Spec](https://docs.google.com/document/d/1xfOaDsV0gllRwZM2zWigc4BEHarDjPKuwbGYp50-Xt0/edit#heading=h.5flr9o646v4k)

[Previous attempt](https://github.com/dimagi/commcare-hq/pull/23183)

Feature flag: MANAGE_RELEASES_PER_LOCATION

this would deprecate in near future
1. toggle RELEASE_BUILDS_PER_PROFILE
2. `corehq.apps.app_manager.models.LatestEnabledBuildProfiles`

best review commit/commit

This is almost ready barring some view beautifications which i can take in a follow up PR as well since this would go through QA.
Dived into some new area in this PR, so might have some naive code.

mkangia committed:
Add new model

mkangia committed:
Add new view

mkangia committed:
add views for update

mkangia committed:
add form and page

mkangia committed:
add util to fetch valid build for a location

mkangia committed:
return enabled build for user

mkangia committed:
add migration

mkangia committed:
add new toggle

mkangia committed:
add specifc exception

mkangia committed:
retturn error if no username i.e not logged in

mkangia committed:
don't check if user assigned no location

mkangia commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,109 @@
+                    url: self.requestUrl(),
+                    success: function (data) {
+                        if (data.success) {
+                            debugger
```
removed later

mkangia committed:
remove redundant imports and ambituous locationutils

mkangia committed:
enable select2 on location id

mkangia committed:
add check to ensure no higher version is enabled for a location

mkangia committed:
use model validation itself

mkangia commented on corehq/apps/domain/forms.py:
```diff
@@ -2454,11 +2455,19 @@ def clean_version(self):
                 self.version_build_id
             except BuildNotFoundException as e:
                 self.add_error('version', e)
+        enabled_release = get_latest_enabled_app_release(self.domain, location_id, app_id)
```
check from here is later removed and only the model validation is reused, in the following commit.

mkangia committed:
lint

mkangia committed:
lint

proteusvacuum commented on corehq/apps/app_manager/models.py:
```diff
@@ -5947,7 +5951,101 @@ def save(self, *args, **kwargs):
+    @classmethod
+    def to_json(cls, domain, location_id, app_id, version):
+        restrictions = cls.objects.filter(domain=domain)
+        if location_id:
```
If these arguments aren't required, I'd make them optional by using kwargs.
I think a nicer way of doing this is to use a django serializer, which you can pass a queryset: https://docs.djangoproject.com/en/2.1/topics/serialization/#serialization-formats-json
That way you can let the caller do other types of filtering without having to modify this method.

mkangia replied:
updated to use kwargs.
I tried using both both DjangoJSONSerializer and ModelSerializer. I also had the challenge of passing `app_names` to the serializer which i could do in `ModelSerializer` but had other challenges. Going to leave this as-is but happy to spend more time if you can suggest how to make changes for this.

proteusvacuum replied:
sorry, that wasn't totally clear. I meant using key word arguments for the optional arguments. something like:
```py
def to_json(cls, domain, location_id=None, app_id=None, version=None):
    ...
```

However, there's something that bugs me about this method, since I would expect a `to_json` method to serialize a single instance, and I would expect to be able to serialize a whole queryset instead of having to pass in options to the method itself. 
I think I would probably have called it like this:
```py
        enabled_app_releases: [
            release.to_json() for release in
            LatestEnabledAppReleases.objects.filter(
                domain=self.domain,
                location-id=self.request.GET.get('location_id'),
                app_id=self.request.GET.get('app_id'), 
                version=self.request.GET.get('version'))
        ]
```

I think that's more explicit, is less code, and gives greater flexibility down the line.

mkangia replied:
[commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/daea93e98f4742e83ebb49a26fc4b8eab7e4ede7)

proteusvacuum commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,104 @@
+            self.activatedOn = ko.observable(details.activated_on);
+            self.deactivatedOn = ko.observable(details.deactivated_on);
+            self.errorMessage = ko.observable();
+            self.dom_id = "restriction_" + self.id;
```
nit this should be `camelCase`

proteusvacuum commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,104 @@
+            self.dom_id = "restriction_" + self.id;
+            self.ajaxInProgress = ko.observable(false);
+            self.actionText = ko.computed(function () {
+                return (self.active() ? "Remove" : "Add");
```
Translate

proteusvacuum commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,104 @@
+    'knockout',
+    'underscore',
+    'hqwebapp/js/initial_page_data',
+    'hqwebapp/js/widgets_v4',
```
nit: we have an unwritten convention to add a comment stating what widget you are using from this import, would be useful to add here.

proteusvacuum commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,62 @@
+                            </tr>
+                        </thead>
+                        <tbody data-bind="foreach: enabledAppReleases, visible: enabledAppReleases">
+                            <tr data-bind="attr: {id: dom_id}, style: { background: rowBgColor() } ">
```
nit: We try not to have inline styles. Would be good to switch the `background` to be a class

proteusvacuum commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,62 @@
+
+{% block page_content %}
+    {% initial_page_data 'enabled_app_releases' enabled_app_releases %}
+    {% registerurl 'location_search' domain %}
```
is this URL used anywhere?

mkangia replied:
was trying to add location search but could not. Would do in a follow up PR.
Removing this. thanks,

proteusvacuum commented on corehq/apps/app_manager/decorators.py:
```diff
@@ -66,13 +68,22 @@ def _safe_cached_download(request, *args, **kwargs):
+        if latest and toggles.MANAGE_RELEASES_PER_LOCATION.enabled(domain):
+            if not username:
+                content_response = dict(error="app.update.not.allowed.user.logged_out",
+                                        default_response="Please log in the app to check for update.")
```
translate. Also, I'd change this to "Please log in _to_ the app to check for _an_ update."

proteusvacuum commented on corehq/apps/app_manager/models.py:
```diff
@@ -5947,7 +5951,101 @@ def save(self, *args, **kwargs):
+    def clean(self):
+        enabled_release = get_latest_enabled_app_release(self.domain, self.location.location_id, self.app_id)
+        if enabled_release:
+            if enabled_release.version > self.version:
```
you could remove a level of indentation here by using an `and`

proteusvacuum commented on corehq/apps/app_manager/models.py:
```diff
@@ -5947,7 +5951,101 @@ def save(self, *args, **kwargs):
+        :param status: expected status
+        """
+        try:
+            enabled_app_release = LatestEnabledAppReleases.objects.get(
```
you could use `get_or_create` with `defaults` here

mkangia replied:
I was using get_or_create here but then updated to be able to use model validation itself in this [commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/e61920e2328af091667f47d78408af01af3546ea)

this also helps me avoid a create and then update, though one can say that that scenario can be avoided by adding `auto_now_add` to `activated_on` because the view does not allow setting active=False and hence the default value is picked always(as of now). 
But then the custom validation i have added would not run as a part of `get_or_create` and i would need to handle that case now after creation. Hence chose to do it this way.

proteusvacuum commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,62 @@
+        <div class="col-sm-12">
+            <div class="panel panel-modern-gray panel-form-only">
+                <div class="panel-body">
+                    <form class="form-horizontal disable-on-submit" id="manage_app_releases" action="" method='post'>
```
nit: html ids normally have `-` instead of `_`

proteusvacuum commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,104 @@
+        function manageReleaseSearchViewModel() {
+            var self = {};
+            self.search = function () {
+                var appId = $("#app_id_search_select").val();
```
Since this is a knockout model, you could probably make these observables and have them prepopulate with the url search string.

mkangia replied:
sorry this is not clear. 
Here, they are trying to fetch the value selected by the user to pass it into the url later.

proteusvacuum replied:
I don't really know what this page looks like, but if the user has done a search, does it make sense to have the search field populated with the search they just carried out? Can you add a screenshot to the PR description?

mkangia replied:
just added a short video.
Those fields are populated back after search via the form

proteusvacuum reviewed (commented):
Left a few nit comments.

mkangia committed:
rename model

mkangia committed:
better comment

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -148,6 +150,7 @@
     module_offers_search,
     get_latest_enabled_build_for_profile,
     get_latest_enabled_versions_per_profile,
+    get_latest_enabled_app_release,
```
Sorry about that.. i already had all the changes, broke it down in different commits later so this was not perfect.

mkangia committed:
raise implicit exception instead

mkangia committed:
fetch ids from query itself and rename

mkangia commented on corehq/apps/app_manager/util.py:
```diff
@@ -638,6 +639,30 @@ def get_latest_enabled_build_for_profile(domain, profile_id):
+            location_id__in=location_and_hierarchy_ids, app_id=app_id, domain=domain, active=True)
+    }
+    for loc_id in location_and_hierarchy_ids:
+        build_id = latest_enabled_releases.get(loc_id)
```
`build_id` in the object won't ever be blank. this checks if there is an entry for this loc_id.

mkangia replied:
I see what you meant.. but i think this needs to be done differently.. following up in a commit later.

mkangia committed:
clearer argument

mkangia committed:
translate

mkangia committed:
camelize

mkangia committed:
add comment for widget being used

mkangia committed:
remove unused url

mkangia committed:
translate and update message

mkangia committed:
compact

mkangia committed:
update ids to be - and not _

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -5947,7 +5950,86 @@ def save(self, *args, **kwargs):
+    deactivated_on = models.DateTimeField(null=True)
+
+    @staticmethod
+    def expire_cache(domain, location_id, app_id):
```
Are you implying that this is better as a util method than being in the model?
That would give me motivation to move this to util and not because its being used in an util method.

proteusvacuum replied:
I don't know why this is part of the model at all. In my mind, the purpose of a staticmethod is to group functions related to a particular class but that don't need an instance (self) or the class (cls). This method acts on a function method that is far away from this class (even though that util method uses the class). 
I would either:
- move that util into this class as a staticmethod
- move this method to be next to the function that it acts on

mkangia replied:
Did the following,
[moved logic to util](https://github.com/dimagi/commcare-hq/pull/23680/commits/6ef2bebb701a67161c1af9997a500556b0bf5dff)
[used signal to address cache expire](https://github.com/dimagi/commcare-hq/pull/23680/commits/6ef2bebb701a67161c1af9997a500556b0bf5dff)

mkangia committed:
fix changes that broke things

mkangia committed:
use kwargs

mkangia committed:
lint

mkangia committed:
use class instead

mkangia committed:
allow migration to from old feature

mkangia committed:
handle case of multiple enabled versions

mkangia committed:
remove duplicate field

proteusvacuum commented on corehq/apps/app_manager/models.py:
```diff
@@ -5996,9 +5996,17 @@ def update_status(cls, domain, app_id, build_id, location_id, version, status):
+            enabled_app_release = LatestEnabledAppReleases(
+                domain=domain, app_id=app_id, build_id=build_id, location_id=location_id, version=version
+            )
+        enabled_app_release.active = status
```
I would call the argument to this method `active` instead of `status`, since that is the field you are setting. `status=True` doesn't really indicate anything.

proteusvacuum commented on corehq/apps/app_manager/util.py:
```diff
@@ -638,6 +639,32 @@ def get_latest_enabled_build_for_profile(domain, profile_id):
+    # get all active enabled releases and order by version desc to get one with the highest version in the end
+    # for a location. Do not use the first object itself in order to respect the location hierarchy and use
+    # the closest location to determine the valid active release
+    latest_enabled_releases = {
```
if you are relying on dict ordering here, you should use an `OrderedDict` (until we switch to Python 3.7+)

mkangia committed:
use arg name same as model

mkangia committed:
use signal to expire cache instead

mkangia committed:
move expire logic to util close to the method

orangejenny commented:
Are there docs on this? Either user-facing or dev-facing?

orangejenny commented on corehq/apps/app_manager/signals.py:
```diff
@@ -44,10 +49,26 @@ def expire_latest_enabled_build_profiles(sender, application, **kwargs):
+    latest enabled release would depend on this location
+    """
+    location = SQLLocation.active_objects.get(location_id=latest_enabled_app_release.location_id)
+    location_and_descendats = location.get_descendants(include_self=True)
```
spelling of `location_and_descendats`

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,101 @@
+) {
+    'use strict';
+    $(function () {
+        var enabledAppRelease = function (details) {
```
`details` has a bunch of properties; it'd be good to verify they're the correct properties with [assert_properties](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/assert_properties.js)

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,101 @@
+                        }
+                    },
+                    error: function () {
+                        self.active(oldStatus);
```
No error message? Is the user going to be able to tell there was an error?

orangejenny replied:
Resolved, I see the generic message is shown when `error` is true.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,101 @@
+            self.toggleStatus = function () {
+                self.active(!self.active());
+            };
+            self.error = ko.observable();
```
I'd combine `error` and `errorMessage` - if there's no message, there's presumably no error - rather than keep track of two separate items.

orangejenny replied:
Having read further, I see you use `error` to control the display of a generic message. I don't think it's needed to show both the generic and specific messages but I don't feel strongly about it.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,101 @@
+            self.enabledAppReleases = ko.observableArray(enabledAppReleases);
+            return self;
+        }
+        var enabledAppReleases = _.map(initialPageData.get('enabled_app_releases'), function (appRelease) {
```
Can shorten this to `_.map(initialPageData.get('enabled_app_releases'), enabledAppRelease)`

mkangia replied:
interesting, thanks

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,101 @@
+        }
+        function manageReleaseSearchViewModel() {
+            var self = {};
+            self.search = function () {
```
It's unusual and a bit confusing to be mixing jQuery manipulation and knockout. Can't you create a `searchValue` observable and then set `data_bind='value: searchValue'` when you add the `app-id-search-select` to the crispy layout? And then similarly for the version and location id?

mkangia replied:
made the change but that didn't work. Reverted it for now, will try to add it back.

orangejenny replied:
Cool, let me know if you want a second set of eyes on it.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,101 @@
+                var appId = $("#app-id-search-select").val();
+                var locationId = $("#location-search-select").val();
+                var version = $("#version-input").val();
+                window.location.search = ("location_id=" + locationId + "&app_id=" + appId + "&version=" +
```
This forces the page to reload with these GET parameters, right? Any reason not to set this up as a form instead?

mkangia replied:
you mean add a new form?
i would like to avoid doing this JS hack if possible while keeping a single form.

orangejenny replied:
I do mean a new form, curious why you'd rather not add one? I think it could be an HTML form - just doing things like wrapping the HTML inputs in a `<form data-bind="submit: search">` and then using [serialize](https://api.jquery.com/serialize/) - not a full Django-based form.

mkangia replied:
i like how this single form is doing both. Don't mind adding one if that adds a value/clarity

orangejenny replied:
Eh, I don't feel strongly about it.

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,101 @@
+            return self;
+        };
+
+        function manageReleasesViewModel(enabledAppReleases) {
```
I don't think this model is adding much value. Curious if you considered just calling koApplyBindings on each individual enabled app release and not having this model exist.

mkangia replied:
i just thought of having a collections view. Agreed that its not adding much value.

> calling koApplyBindings on each individual enabled app release
How do you suggest i would do that?

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+            self.fields['location_id'].initial = request.GET.get('location_id')
+        if request.GET.get('version'):
+            self.fields['version'].initial = request.GET.get('version')
+        self.helper = FormHelper()
```
Use [HQFormHelper](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/crispy.py#L21) here to get standard HQ proportions. (It's new.)

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+        self.helper.form_tag = False
+
+        self.helper.layout = crispy.Layout(
+            crispy.Field('app_id', id='app-id-search-select'),
```
This should also be `ko-select2`, so all the selects on the page match. In general plain select boxes aren't great - long lists should use select2s and short lists should use button toggles.

mkangia replied:
button toggles ?

orangejenny replied:
these toggles: https://www.commcarehq.org/styleguide/molecules/#molecules-selections

mkangia replied:
aha! okay. I expect this to be long enough to keep it as select.

orangejenny replied:
agreed

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+        self.helper.layout = crispy.Layout(
+            crispy.Field('app_id', id='app-id-search-select'),
+            crispy.Field('location_id', id='location-search-select', css_class="ko-select2"),
+            crispy.Field('version', id='version-input'),
```
Can this use a select2 dropdown and show the version number and comment?

This commit sets up a version select2 that changes its options based on the value of an application dropdown: https://github.com/dimagi/commcare-hq/pull/23521/commits/47e80043d0ba461c6276fce3d8ad134d17ea2d88

mkangia replied:
Thanks for suggesting that, i see that that js change now returns [build_id instead of version](https://github.com/dimagi/commcare-hq/pull/23521/files#diff-62b3073366d0a0237a9467da2df883b4R31). This is good. I can remove some logic here to find out build from version and also validation around it.
Though i think we broke the implementation in the PR you linked coz it still relies on getting a version number.

mkangia replied:
while making the changes i realised it was better to just update the parent js to use version instead of build id as the key since we need to pass on version anyway so we would need both build id and version.. plus the other place where this was changed for `AppTranslationForm` also needs version and not build id.

mkangia replied:
[commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/f6bf3a72606fa4f664b72efa178f330c35d51829)

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+        )
+
+    def app_id_choices(self):
+        choices = ([(None, 'Select Application')])
```
translate?

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+        return choices
+
+    def location_id_choices(self):
+        choices = ([(None, 'Select Location')])
```
translate?

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+    def location_id_choices(self):
+        choices = ([(None, 'Select Location')])
+        for location in SQLLocation.active_objects.filter(domain=self.domain):
+            choices.append((location.location_id, location.name))
```
Should these be sorted? Are they already sorted?

mkangia replied:
no need to sort them.. user would be typing and selecting via select2 only.

orangejenny replied:
The user is still going to see the list, it'd be good for that list to look sensical. Perhaps not worthwhile if it's expensive to sort, but if it's cheap, might as well do it.

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+            try:
+                self.version_build_id
+            except BuildNotFoundException as e:
+                self.add_error('version', e)
```
Shouldn't this be `six.text_type(e)`?

mkangia replied:
i think Form internally handles that.

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+            LatestEnabledAppRelease.update_status(self.domain, app_id, self.version_build_id, location_id,
+                                                  version, active=True)
+        except ValidationError as e:
+            return False, ','.join(e.messages)
```
Nitpick: `','` => `', '` if humans are going to read this.

kaapstorm replied:
Exceptions don't have a `messages` attribute in Python 3. Best to use `six.text_type(e)` instead.

kaapstorm replied:
My mistake. It's a ValidationError, so it does have a `messages` attribute.

orangejenny commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,64 @@
+        <div class="col-sm-12">
+            <div class="panel panel-modern-gray panel-form-only">
+                <div class="panel-body">
+                    <form class="form-horizontal disable-on-submit" id="manage-app-releases" action=""
```
Huh, I didn't know `disable-on-submit` worked on forms, I thought it had to be applied to the buttons themselves. Cool.

orangejenny commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,64 @@
+                        {% crispy manage_releases_form %}
+                    </form>
+                {% if enabled_app_releases %}
+                    <table id="managed-releases" class="table">
```
Let's add `table-striped` too, it makes long tables more readable.

mkangia replied:
`table-striped` conflicted with the bg color i was adding according to status so i didn't use that.

orangejenny replied:
Doesn't the green just override the table-striped color? That seems fine - mostly alternating rows, but possibly with a few green ones at the top.

orangejenny commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,64 @@
+                    <table id="managed-releases" class="table">
+                        <thead>
+                            <tr>
+                                <th>{% trans 'Application' %}</th>
```
It's good practice to explicitly set column proportions by adding a `col-sm-X` class to each <th>. It prevents tables from getting wonky if someone has a really long application or location name.

orangejenny commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,64 @@
+                                <th>{% trans 'Action' %}</th>
+                            </tr>
+                        </thead>
+                        <tbody data-bind="foreach: enabledAppReleases, visible: enabledAppReleases">
```
Should `visible: enabledAppReleases` go on the table tag? I don't think an empty tbody is visible anyway.

mkangia replied:
was just thinking of letting the user know that there is an empty table so we searched but found nothing.
You think a better message would be "No entries found"

orangejenny replied:
Yeah, a message would be nice, makes it more obvious there wasn't an error.

orangejenny commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,64 @@
+                                <td>
+                                    <a href="#" data-bind =
+                                    "click: toggleRestriction, text: actionText, visible: !ajaxInProgress()"></a>
+                                    <span class="spinner fa fa-spinner" data-bind="visible: ajaxInProgress"></span>
```
I don't think the `spinner` class is doing anything. `fa-spin fa fa-refresh` is what we usually use for buttons that are working - the `fa-refresh` makes it the circular arrow image and the `fa-spin` makes it animate.

mkangia replied:
`spinner` is used for handling visibility it during ajax request.
Are you suggesting to use a button instead of a link and then updating classes on it during ajax request?

mkangia replied:
Seeing the following comment where you suggest updating this to a button.

orangejenny commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -0,0 +1,64 @@
+                                    </div>
+                                </td>
+                                <td>
+                                    <a href="#" data-bind =
```
This should be a button, not a link (links are for navigation, buttons are for actions).

orangejenny reviewed (commented):
Only reviewed front-end code, all of my feedback is minor stuff.

kaapstorm commented on corehq/apps/domain/views/releases.py:
```diff
@@ -0,0 +1,103 @@
+        latest_enabled_app_release.full_clean()
+    except ValidationError as e:
+        response_content = {
+            'message': ','.join(e.messages)
```
(ditto `six.text_type(e)`)

kaapstorm replied:
My bad. This is a ValidationError. Resolving.

proteusvacuum commented on corehq/apps/app_manager/models.py:
```diff
@@ -5961,21 +5961,11 @@ class LatestEnabledAppRelease(models.Model):
+    def save(self, *args, **kwargs):
+        super(LatestEnabledAppRelease, self).save(*args, **kwargs)
+        from corehq.apps.app_manager import signals
+        signals.latest_enabled_app_release_post_save.send(LatestEnabledAppRelease,
```
using a signal here is a bit confusing. I think signals are normally meant to be used to decouple things between apps that shouldn't really know anything about each other, but might be interested in the same events.
I would just call `expire_get_latest_enabled_app_release_cache` from here.

mkangia replied:
[commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/6bd9823e890b916e29347c1fd6ba9d86a8b5fe11)

esoergel commented on corehq/apps/domain/views/releases.py:
```diff
@@ -78,13 +79,20 @@ def activate_release_restriction(request, domain, restriction_id):
+        latest_enabled_app_release.full_clean()
+    except ValidationError as e:
+        response_content = {
+            'message': ','.join(e.messages)
```
Seems like this should return a 400 response code or something

mkangia replied:
we don't set `success` and then it's handled at ajax. If that's not a good practice and i should change it to 400 and then change the javascript code instead, don't mind doing that.

esoergel replied:
I don't think it's a huge deal - whichever feels cleaner.  The main advantage of using response codes is that other tools will know what the response means, such as your browser console or the jquery ajax library (it has hooks for `error` and `statusCode` callbacks).  I'd probably default to sticking with standards, but wouldn't be foolishly consistent if it ends up being messier than doing something custom.

esoergel commented on corehq/apps/domain/views/releases.py:
```diff
@@ -49,8 +50,12 @@ def page_context(self):
+                return redirect(self.urlname, self.domain)
+            else:
+                messages.error(request, error_message)
+                return self.get(request, *args, **kwargs)
```
nit: this feels like a messy interface - could you move the validation check to form.clean?  It's a user error, so it'd make sense to raise there.

mkangia replied:
I actually made an effort to move the validation message to model in this [commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/e61920e2328af091667f47d78408af01af3546ea) so that its not duplicated and hence its not a part of form clean.

mkangia replied:
offlined.

esoergel commented on corehq/apps/app_manager/decorators.py:
```diff
@@ -66,13 +69,27 @@ def _safe_cached_download(request, *args, **kwargs):
         latest_enabled_build = None
-        if latest and request.GET.get('profile') and toggles.RELEASE_BUILDS_PER_PROFILE.enabled(domain):
-            latest_enabled_build = get_latest_enabled_build_for_profile(domain, request.GET.get('profile'))
+        if latest and toggles.MANAGE_RELEASES_PER_LOCATION.enabled(domain):
```
There's a lot of code here about how to get the correct app.  I'd pull it out to a standalone function so it's easier to understand.  That'd also be a good place to test.

esoergel commented on corehq/apps/app_manager/util.py:
```diff
@@ -649,11 +649,15 @@ def get_latest_enabled_app_release(domain, location_id, app_id):
         LatestEnabledAppRelease.objects.filter(
-            location_id__in=location_and_ancestor_ids, app_id=app_id, domain=domain, active=True)
+            location_id__in=location_and_ancestor_ids, app_id=app_id, domain=domain, active=True).order_by(
+            'version')
```
You might be able to cut down results returned here with a `.distinct()`, though I'm not certain what that'd look like

mkangia replied:
not sure how that would help. we need all values set for a location.

esoergel replied:
what for?  Don't you only need the latest one?  The dict comprehension only preserves one value per location, the others are thrown away.

mkangia replied:
offlined. distinct would still keep all since version would be different.

esoergel reviewed (commented):
I had a lot of nitpicky comments (many of which were addressed in subsequent commits), but at a high level, it looks good.

I would've liked to see the commits organized a bit differently for ease of review - the main commits at the beginning were broken up by file, which meant that you'd see new code that referenced stuff that wasn't yet added, so I had to jump between commits to make sense of it.  I'd prefer a gradual expansion of all those files in parallel, so each commit is functional - one commit with the boilerplate of setting up a new page, the next adding the relevant fields, one for validation, one for styling, etcetera.

mkangia committed:
improvements by review feedback

mkangia committed:
avoid using signal for expire cache

mkangia committed:
update util js to use version instead of id and then use that for version

mkangia commented:
@orangejenny thanks for the review,
addressed all review in this [commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/402183e22d9ed4c87e0c79f05064d9e9d76b89d6) or left comments for others

mkangia committed:
remove unused imports

mkangia committed:
extract method

mkangia commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2383,91 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+    def clean_version(self):
+        if not self.cleaned_data.get('version'):
+            self.add_error('version', _("Please enter version"))
+        return self.cleaned_data.get('version')
```
The same form also handles search for which these are not required. It started showing error messages we re-render the form after search so i had to make this change.

mkangia commented on corehq/apps/app_manager/decorators.py:
```diff
@@ -73,7 +74,11 @@ def _safe_cached_download(request, *args, **kwargs):
+            if not username:
+                content_response = dict(error="app.update.not.allowed.user.logged_out",
+                                        default_response="Please log in the app to check for update.")
+                return HttpResponse(status=406, content=json.dumps(content_response))
```
we check for `latest` bool in this which is set only when someone is requesting for an `update` so the initial app installation should proceed without this check.

esoergel replied:
Okay.  Still seems scary to me - I'd check with mobile unless you're confident that that's okay.

mkangia replied:
Sure, i will test this

mkangia replied:
tested that the bool `latest` is set only in case of checking for update.
I installed a version using application scan barcode successfully. Didn't test ccz because thats an offline install anyway.

mkangia commented on corehq/apps/app_manager/util.py:
```diff
@@ -638,6 +639,30 @@ def get_latest_enabled_build_for_profile(domain, profile_id):
+        enabled_release.location_id: enabled_release.build_id
+        for enabled_release in
+        LatestEnabledAppReleases.objects.filter(
+            location_id__in=location_and_hierarchy_ids, app_id=app_id, domain=domain, active=True)
```
this logic was later updated to order by version. Hence eventually we have the latest one for each location after iterating.

mkangia commented on corehq/apps/app_manager/util.py:
```diff
@@ -638,6 +639,30 @@ def get_latest_enabled_build_for_profile(domain, profile_id):
+    for loc_id in location_and_hierarchy_ids:
+        build_id = latest_enabled_releases.get(loc_id)
+        if build_id:
+            return get_app(domain, build_id)
```
i think i am fine with this as-is. If i also return the latest build it adds 
1. more logic/responsibility to this method
2. plus now would need to handle expiry of this method each time a build is made which can have a complex logic of its own.

esoergel replied:
that makes sense - maybe a wrapper function that calls both would make more sense then?  My main point is that it seems like there should be one function whose responsibility is to get the appropriate app for any user on a domain that uses this feature flag, including codifying the expected behavior if no release for their location is found.

mkangia replied:
offlined and concluded to leave it as-is in interest of not repeating the logic to get the latest app at both places - with or without the FF?

mkangia commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2382,87 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+
+    def location_id_choices(self):
+        choices = ([(None, 'Select Location')])
+        for location in SQLLocation.active_objects.filter(domain=self.domain):
```
i was trying to using location search here but i failed. So i am going to do that in a following PR or the same one while QA is on.
adding values_list

mkangia commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2382,87 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+        )
+
+    def app_id_choices(self):
+        choices = ([(None, 'Select Application')])
```
hmm, i thought they always needed a tuple.

esoergel replied:
I think it needs a list of tuples, but this isn't a tuple unless you put a trailing parenthesis in:
```python
In [1]: ([]) == []
Out[1]: True

In [2]: ([],) == []
Out[2]: False
```

mkangia commented on corehq/apps/domain/views/releases.py:
```diff
@@ -48,3 +48,38 @@ def post(self, request, *args, **kwargs):
+        'deactivated_on': (datetime.strftime(latest_enabled_app_release.deactivated_on, '%Y-%m-%d %H:%M:%S')
+                           if latest_enabled_app_release.deactivated_on else None),
+    }
+    return HttpResponse(json.dumps(response_content), content_type='application/json')
```
i like having two separate urls. Extracting out logic though.

mkangia commented on corehq/apps/app_manager/models.py:
```diff
@@ -5947,7 +5950,86 @@ def save(self, *args, **kwargs):
         super(GlobalAppConfig, self).save(*args, **kwargs)
 
 
+class LatestEnabledAppReleases(models.Model):
```
`AppReleaseSettings` , `AppReleaseRestriction`, `AppReleaseByLocation` ?

mkangia commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2382,87 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+        version = self.cleaned_data['version']
+        app_id = self.cleaned_data['app_id']
+        LatestEnabledAppReleases.update_status(self.domain, app_id, self.version_build_id, location_id, version,
+                                               True)
```
this was updated in another review comment.

mkangia committed:
improvements in review feedback

mkangia commented:
Thanks for the review @esoergel ,
addressed 
https://github.com/dimagi/commcare-hq/pull/23680/commits/852f37c84424a4125da7ad2c179b52602b0b6090
and
https://github.com/dimagi/commcare-hq/pull/23680/commits/d5242a4cb8206a29906ed1ae28acaaffd21a6cb4

mkangia committed:
fix review change

mkangia committed:
update to_json

mkangia committed:
fix filter

mkangia committed:
revert changes that broke js

mkangia committed:
validate only when active

mkangia committed:
extract logic for 406

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/manage_releases.js:
```diff
@@ -0,0 +1,105 @@
+    ko,
+    _,
+    initialPageData,
+) {
```
Noting that this lint error is important; the extra comma will break our minification. It won't block deploy, but the file won't get minified.

orangejenny commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2373,90 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+        if request.GET.get('version'):
+            self.fields['version'].initial = request.GET.get('version')
+        self.helper = HQFormHelper()
+        self.helper.label_class = 'col-sm-3 col-md-2'
```
Using `HQFormHelper` means you can delete the `label_class` and `field_class` lines.

orangejenny commented on corehq/apps/domain/templates/domain/manage_releases.html:
```diff
@@ -44,9 +44,11 @@
-                                    <a href="#" data-bind =
-                                    "click: toggleRestriction, text: actionText, visible: !ajaxInProgress()"></a>
-                                    <span class="spinner fa fa-spinner" data-bind="visible: ajaxInProgress"></span>
+                                    <button type="button" class="" data-bind="click: toggleRestriction">
```
This is unstyled, it should use `btn` and `default-btn`.
<img width="204" alt="Screen Shot 2019-03-25 at 10 10 03 AM" src="https://user-images.githubusercontent.com/1486591/54930965-30390a80-4ee6-11e9-8c1b-ef918ea5636c.png">

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/widgets_v4.js:
```diff
@@ -28,7 +28,7 @@ hqDefine("app_manager/js/widgets_v4", [
                         results: _.map(data.apps, function (build) {
                             return {
-                                id: build.id,
+                                id: build.version,
```
This breaks the build comparison on the view source files page, which is still expecting a build id.

mkangia replied:
added, https://github.com/dimagi/commcare-hq/pull/23771

mkangia committed:
Merge branch 'master' into mk/release-location

mkangia committed:
rename model

mkangia committed:
fix logic to let update handle full clean

mkangia committed:
fix logic to fetch latest enabled build

mkangia committed:
review improvements

mkangia committed:
disable button and add class according to action

mkangia committed:
lint

orangejenny commented:
@mkangia Thanks for the updates, all of my feedback is addressed.

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into mk/release-location

mkangia commented:
thanks for mater update @esoergel.

esoergel commented on corehq/apps/app_manager/models.py:
```diff
@@ -5950,7 +5950,7 @@ def save(self, *args, **kwargs):
 
 
-class LatestEnabledAppRelease(models.Model):
+class AppReleaseByLocation(models.Model):
```
I like this change overall, although I think for most of these the context is clear enough that you could refer to an individual entry simply as `release`, following the principle "Add no more context to a name than is necessary".  That is, it may be useful to clarify what release we're talking about in functions that are called externally, but not necessarily within such a function or for semi-private functions.

mkangia replied:
Makes sense, [Renamed](https://github.com/dimagi/commcare-hq/pull/23680/commits/7ad06f401d3cdc636fdd91004ef68b94b1ae5950)

esoergel commented on corehq/apps/app_manager/util.py:
```diff
@@ -639,12 +639,12 @@ def get_latest_enabled_build_for_profile(domain, profile_id):
 
 @quickcache(['domain', 'location_id', 'app_id'], timeout=24 * 60 * 60)
-def get_latest_enabled_app_release(domain, location_id, app_id):
+def get_app_release_by_location(domain, location_id, app_id):
```
This one actually is the latest enabled release, right?

mkangia replied:
[renamed](https://github.com/dimagi/commcare-hq/pull/23680/commits/a1bc83cc1f244c5f4e4acf0e5b20f0b7ac8934e1)

esoergel commented on corehq/apps/domain/forms.py:
```diff
@@ -2377,3 +2373,88 @@ def __init__(self, defaults=None, disable_input=False, **kwargs):
+    def location_id_choices(self):
+        choices = [(None, _('Select Location'))]
+        choices.extend(SQLLocation.active_objects.filter(domain=self.domain).values_list('location_id', 'name'))
+        return choices
```
`name` is not unique on a domain, so users won't be able to distinguish.  Elsewhere we use `location.display_name`, which includes the type, but my preference is `get_path_display`, which is guaranteed to be unique (each location name is unique among immediate siblings).

@orangejenny wrote
> The user is still going to see the list, it'd be good for that list to look sensical. Perhaps not worthwhile if it's expensive to sort, but if it's cheap, might as well do it.

The proper ordering for a path display is "tree order", either depth first or breadth first, not sure which is better.  I believe it happens automatically with a `get_descendants` query.  Probably out of scope of this PR, though.

More critical, however is the volume here.  IIRC, ICDS has (or will have) around 4 million locations, though you could lower that substantially by excluding the bottom couple levels.  That's probably too many to render in the html up-front.  Luckily, we do already have the server side of an ajax select2 widget for locations: https://github.com/dimagi/commcare-hq/blob/9bdfffdaf6f8a5a500cc5c4f9a01f99ce48f799b/corehq/apps/locations/urls.py#L25-L25
It's based on the "Groups or Users" filter options, but only pulls locations.  Might do some weird namespacing thing with the ids though, I don't recall.

mkangia replied:
Thanks @esoergel 
Integrated in [commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/412d18c72dcfe4cf5449bbd9d6f50853a5f054a3). Also updated the library to allow initial value in [commit](https://github.com/dimagi/commcare-hq/pull/23680/commits/f837c589a5a0316be798506052479b56b83e5f91)

mkangia committed:
Merge branch 'master' into mk/release-location

mkangia committed:
show complete path in list

mkangia committed:
inject initial value to app version search

mkangia committed:
update location util to accept initial value

mkangia committed:
integrate location search

mkangia committed:
remove redundant version assignment

mkangia committed:
lint

mkangia committed:
rename

mkangia committed:
remove unused import

mkangia committed:
lint

mkangia committed:
rename to exclude extra context

mkangia commented:
@esoergel @orangejenny 
Can you take a look at the updates post last review?

Also i have close to 200 comments but can i get a "approved"/ 👍 if this is good to go according to code review?

esoergel reviewed (approved):
It's been a while since I did the full review of this, but I don't think I have any unaddressed feedback.

orangejenny commented:
Looks good to me - I only ever looked at the front end code, but between me and Ethan I think we've got everything covered.

mkangia committed:
only show released builds for selecting app version

mkangia committed:
add model validation to allow only released builds

mkangia committed:
dont allow unrelease on a build if restrictions added

mkangia committed:
lint

orangejenny commented on corehq/apps/app_manager/views/releases.py:
```diff
@@ -230,8 +234,10 @@ def current_app_version(request, domain, app_id):
-            return json_response({'error': _('Please disable any enabled profiles to un-release this build.')})
+        if (LatestEnabledBuildProfiles.objects.filter(build_id=saved_app_id).exists() or
+                AppReleaseByLocation.objects.filter(build_id=saved_app_id, active=True).exists()):
+            return json_response({'error': _('Please disable any enabled profiles/location restriction '
```
Didn't notice this earlier: does this error message actually get displayed to the user?

mkangia replied:
Ya it was added after your last review. This things were suggested recently.
Yes, this get displayed to the user in browser prompt..

orangejenny replied:
Cool. Browser popups are not totally ideal, but they're quick to implement, so that seems fine.

mkangia replied:
Agreed. I see that i added this last time considering how rarely this would be used and didn't pay much effort to it. Will revisit that.

orangejenny commented:
Last few commits look fine.

mkangia committed:
merge master

mkangia committed:
Merge branch 'master' into mk/release-location

mkangia commented:
Here it goes, thanks for tight review all.
